src/System.Text.Http/System/Text/Http/Parser/HttpParser.cs

```diff
@@ -44,714 +41,310 @@ public HttpParser(bool showErrorDetails)
+            examined = buffer.End;
+            var reader = new BufferReader<byte>(buffer);
+            bool success = ParseRequestLine(handler, ref reader);
+            consumed = reader.Position;
+            if (success) examined = reader.Position;
+            return success;
+        }
+
+        public bool ParseRequestLine<T>(T handler, in ReadOnlySequence<byte> buffer, out int consumed) where T : IHttpRequestLineHandler
+        {
+            // Look for CR/LF
+
+            BufferReader<byte> reader = new BufferReader<byte>(buffer);
+            if (!reader.TryReadToAny(out ReadOnlySpan<byte> requestLine, Eol, advancePastDelimiter: false))
             {
-                consumed = lineIndex + 1;
-                span = span.Slice(0, lineIndex + 1);
+                // Couldn't find a delimiter
+                consumed = 0;
+                return false;
             }
-            else
+
+            if (!reader.IsNext(Eol, advancePast: true))
             {
-                var eol = Sequence.IndexOf(buffer, s_Eol[0], s_Eol[1]);
-                if (eol == -1)
+                if (reader.TryRead(out byte value) && value == ByteCR && !reader.TryRead(out value))
                 {
+                    // Incomplete if ends in CR
                     consumed = 0;
                     return false;
                 }
 
-                span = PipelineExtensions.ToSpan(buffer.Slice(0, eol + s_Eol.Length));
+                // Not CR/LF
+                RejectRequestLine(requestLine);
             }
 
-            // Fix and parse the span
-            fixed (byte* data = &MemoryMarshal.GetReference(span))
-            {
-                ParseRequestLine(ref handler, data, span.Length);
-            }
+            ParseRequestLine(handler, requestLine);
 
-            consumed = span.Length;
+            consumed = (int)reader.Consumed;
             return true;
         }
 
-        private unsafe void ParseRequestLine<T>(ref T handler, byte* data, int length) where T : IHttpRequestLineHandler
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ParseRequestLine<T>(T handler, in ReadOnlySpan<byte> data) where T : IHttpRequestLineHandler
         {
-            // Get Method and set the offset
-            var method = HttpUtilities.GetKnownMethod(data, length, out var offset);
-
-            ReadOnlySpan<byte> customMethod =
-                method == Http.Method.Custom ?
-                    GetUnknownMethod(data, length, out offset) :
-                    default;
+            // The absolute smallest request line is something like "Z * HTTP/1.1"
+            // See https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2
 
-            // Skip space
-            offset++;
-
-            byte ch = 0;
-            // Target = Path and Query
-            var pathEncoded = false;
-            var pathStart = -1;
-            for (; offset < length; offset++)
+            if (data.Length < 11)
```

Comment:
`Z * HTTP/1.1` is 12 characters. Why is the check < 11?

Also does this hold true for the newer RFC?
https://tools.ietf.org/html/rfc7230#section-5.3.4

Reply:
> Z * HTTP/1.1 is 12 characters. Why is the check < 11?

Because I can't count. The real thing we need here is <9 to make sure we can read our known headers out (long and a space).

> Also does this hold true for the newer RFC?

That example is longer. This is just a slight optimization to avoid checking twice for header length when reading out known headers. I'll comment.